### DISCUSSION TITLE
Prisma Custom Client Location

### DISCUSSION THREAD

**Discussion by jd4u** [notiz-dev/nestjs-prisma#55] [Q&A] [answered]
[Custom Prisma Client Location](https://nestjs-prisma.dev/docs/custom-prisma-client-location/)

Is it required to CustomPrismaModule/CustomPrismaService? 
or 
using the appropriate location with PrismaModule /PrismaService is sufficient?

**Comment by jd4u** [comment, marked as answer]
After exploring more, I see the following are required when prisma custom client location is used. Please correct me if wrong anywhere...

1] Use of CustomPrismaModule/CustomPrismaService (not to use PrismaModule/PrismaService anywhere)

2] Use Custom Location for referencing PrismaClient (not to use @prisma/client anywhere)
```
import { PrismaClient } from '@myLib/prisma'; // 👈 update to your output directory
```
3] Use CustomPrismaModule to register unique PrismaClient
```
CustomPrismaModule.forRoot({
      name: 'custPrismaService', // 👈 must be unique for each PrismaClient
      client: new PrismaClient(), // create new instance of PrismaClient
}),
```
4] Use for Injection CustomPrismaService<PrismaClient> (not to use PrismaService from @prisma/client)
```
Inject('custPrismaService') private readonly cpsService: CustomPrismaService<PrismaClient>; 
// Use ☝️ the unique name for PrismaClient 

```
5] Use prisma.client instead of prisma to interact with db. (or create another local variable prisma and assign cpsService.client to it.
```
private prisma: PrismaClient;
constructor(Inject('PrismaService') private readonly cpsService: CustomPrismaService<PrismaClient>){
   this.prisma = cpsService.client;
```
6] In test, use `dbServ = await app.get<CustomPrismaService<PrismaClient>>('custPrismaService')`; // 👈 must be the unique name of PrismaClient
[ not to use `dbServ = await app.get(CustomPrismaService<PrismaClient>);`  // 👈 this will give error saying 'Nest could not find CustomPrismaService element (this provider does not exists in current context) ' ]

**Reply by jd4u** [reply]
Though I marked the above post as answer, please do correct where ever should be...

**Reply by marcjulian** [reply]
Hi @jd4u, yes you are correct if you have a custom location for the Prisma Client you need to use `CustomPrismaModule` and respectively the `CustomPrismaService` to use the Prisma Client.

To inject the `CustomPrismaService` you will need to use the unique name which you provided in the forRoot of `CustomPrismaModule`. In your example you use the name `custPrismaService`, so you need to use it in `Inject` too:

```diff
- Inject('PrismaService') private readonly cpsService: CustomPrismaService<PrismaClient>
+ Inject('custPrismaService') private readonly cpsService: CustomPrismaService<PrismaClient>
```

Do you think the docs for [Custom Prisma Client Location](https://nestjs-prisma.dev/docs/custom-prisma-client-location/) could be improved? Do you have any suggestions?

**Reply by jd4u** [reply]
Yes, the docs need to be improved... that the reason I posted the sequential approach to implement here for others like me!!

**Reply by marcjulian** [reply]
How do you thing it could be improved? Is the only information missing that`CustomPrismaModule`/`CustomPrismaService` must be used when using custom client location? The usage is already described on the doc page.

**Reply by jd4u** [reply]
As theme of the site is like page per topic, covers multiple features of the topic. If something like recipe approach take, in few topics it will help. 

In [Custom Prisma Client Location](https://nestjs-prisma.dev/docs/custom-prisma-client-location/), there are three features described. 1] Nestjs-prisma supports custom output location of generated client. 2] Use CustomPrismaModule with custom output location and 3] Multiple Prisma Clients support by nestjs-prisma.

At first time user of nestjs-prisma, I browsed all pages and took these three aspect different and could not build understanding, I assume that's because of the content theme. As most are covered there, but missing the continuity for understanding purpose.

**Reply by marcjulian** [reply]
Thanks for your feedback, it's always good to view the docs as a first time user. I will think about how to improve the page or maybe ad a recipe/guide about the topic
